search_files.py

```diff
@@ -0,0 +1,152 @@
+
+    data = {"query": query}
+    batches = 1
+    total, search_after = search_files(url, params, data, output_file)
+    while search_after is not None:
+        data["searchAfter"] = quote(json.dumps(search_after))
+        count, search_after = search_files(url, params, data, output_file)
+        batches += 1
+        total += count
+
+    print(f"{total} file(s) found in {batches} batches")
+
+
+def search_files(
+    url: str,
+    params: dict,
+    data: dict,
+    output_file: str,
+) -> "Tuple[int, dict]":
+    """Perform a single search, possibly continuing to paginate after a previous search.
+
+    Args:
+        url (str): URL with path.
+        params (dict): Query parameters.
+        data (dict): Request body.
+        output_file (str): File to write founds filepaths to.
+
+    Raises:
+        RuntimeError: `if response.status_code != 200` 
+
+    Returns:
+        Tuple[int, dict]: Number of results, search_after object.
+    """
+    response = requests.post(url=url, data=data, params=params, verify=VERIFY)
+    if response.status_code != 200:
+        raise RuntimeError(response.text)
+
+    content = json.loads(response.content)
+    results = content["results"]
+    with open(output_file, "a") as f:
+        for result in results:
+            f.write(f"{result['_source']['location']}\n")
+
+    count = len(results)
+    print(f"{count} file(s) found")
+    if "search_after" in content:
+        return count, content["search_after"]
+    else:
+        return count, None
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        prog="search_files",
+        description=(
+            "Performs DataGateway searches for Datafiles matching the provided query."
+            " These paths will be written to file in\nbatches, and can then be "
+            "inspected and filtered further if needed before submitting using "
+            "queue_file_downloads."
+        ),
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+    parser.add_argument(
+        "output_file",
+        type=str,
+        help=(
+            "File to append newline separated paths to.\n"
+            "This can then be provided as an input to queue_file_downloads."
+        ),
+    )
+    parser.add_argument(
+        "query",
+        type=str,
+        help=
+"""Lucene syntax formatted search query. Full help text and examples can be found
+in the DataGateway UI. Note that wildcards can significantly increase the time
+taken to perform a search, and the more specific the search query is the more
+efficient it will be. Some example searches are:
+    'visitId:AB1234'
+        Search for all Datafiles in all parts of proposal
+    'visitId:\"AB1234-1\"'
+        Search for all Datafiles in a (part) visit
+    'location.fileName:txt'
+        Search for Datafiles with the extension 'txt'
+    'location:raw'
+        Search for Datafiles with the directory 'raw' somewhere in their path
```

Comment:
Is this actually looking for 'raw' directories (ie. /raw/ in the location) or just looking for the string 'raw' anywhere in the location? If not directory specific then we should probably change the word directory to string.

Reply:
Technically it would also match `/dls/data/raw` i.e. a file named `raw`. It would not match `/dls/data/brawl/file.txt`, since the token is `brawl` and that is not a match for `raw`.